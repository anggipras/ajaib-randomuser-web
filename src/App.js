import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { 
  Navbar,
  Container,
  Form,
  FormControl,
  Button,
  Dropdown,
  Table
 } from 'react-bootstrap';
import { useEffect, useState } from 'react';
import { faSort } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import axios from 'axios';
import moment from 'moment';
import InfiniteScroll from 'react-infinite-scroll-component';
import { TailSpin } from 'react-loader-spinner'

let sortMethod = 'asc'
function App() {
  const [gender, setGender] = useState('All Gender')
  const [inputData, setInputData] = useState('')
  const [newSort, setNewSort] = useState(false)
  const [hasMore, setHasMore] = useState(false)
  const [page, setPage] = useState(0)
  const [userData, setUserData] = useState([
    {
      username: "",
      name: "",
      email: "",
      gender: "",
      registered_date: "" 
    }
  ])
  const [userSearchData, setUserSearchData] = useState([
    {
      username: "",
      name: "",
      email: "",
      gender: "",
      registered_date: "" 
    }
  ])

  // First Load Page
  useEffect(() => {
    if (JSON.parse(localStorage.getItem('FIRST_LOAD'))) {
      localStorage.removeItem('FIRST_LOAD')
    }

    axios.get(`https://randomuser.me/api/?page=0&results=30`)
    .then((res) => {
      let theResult = res.data.results
      let newArrData = theResult.map(val => {
        let newObject = {
          username: val.login.username,
          name: `${val.name.first} ${val.name.last}`,
          email: val.email,
          gender: val.gender,
          registered_date: val.registered.date
        }
        return newObject
      });
      localStorage.setItem('FIRST_LOAD', JSON.stringify(newArrData))
      setUserData(newArrData)
      setPage(page+1)
      setHasMore(true)
    }).catch(err => console.log(err))
  },[])

  // Fetching Random Data By Pagination
  const fetchRandomDataByPagination = () => {
    axios.get(`https://randomuser.me/api/?page=${page}&results=10`)
    .then((res) => {
      let theResult = res.data.results
      let newArrData = theResult.map(val => {
        let newObject = {
          username: val.login.username,
          name: `${val.name.first} ${val.name.last}`,
          email: val.email,
          gender: val.gender,
          registered_date: val.registered.date
        }
        return newObject
      });
      setTimeout(() => {
        setUserData([...userData].concat(newArrData))
        if (JSON.parse(localStorage.getItem('FIRST_LOAD'))) {
          let localStrg = JSON.parse(localStorage.getItem('FIRST_LOAD'))
          localStorage.setItem('FIRST_LOAD', JSON.stringify([...localStrg].concat(newArrData)))
        }
        setPage(page+1)
      }, 1000);
    }).catch(err => {
      console.log(err);
      setHasMore(false)
    })
  }

  // Sort based on Table Data
  const sortTable = (val) => {
    if (sortMethod === 'asc') {
      sortMethod = 'dsc'
    } else {
      sortMethod = 'asc'
    }

    let newUserSortData = userData
    
    newUserSortData.sort((a, b) => {
      if (sortMethod === 'asc') {
        if (val === "name") {
          return (a.name === b.name) ? 0 : (a.name < b.name) ? -1 : 1
        } else if (val === "email") {
          return (a.email === b.email) ? 0 : (a.email < b.email) ? -1 : 1
        } else if (val === "gender") {
          return (a.gender === b.gender) ? 0 : (a.gender < b.gender) ? -1 : 1
        } else {
          return (a.registered_date === b.registered_date) ? 0 : (a.registered_date < b.registered_date) ? -1 : 1
        }
      } else {
        if (val === "name") {
          return (a.name === b.name) ? 0 : (a.name > b.name) ? -1 : 1
        } else if (val === "email") {
          return (a.email === b.email) ? 0 : (a.email > b.email) ? -1 : 1
        } else if (val === "gender") {
          return (a.gender === b.gender) ? 0 : (a.gender > b.gender) ? -1 : 1
        } else {
          return (a.registered_date === b.registered_date) ? 0 : (a.registered_date > b.registered_date) ? -1 : 1
        }
      }
    })
    setNewSort(!newSort)
    setUserData(newUserSortData);
  }

  // Reset All Filter
  const resetFilter = () => {
    if (JSON.parse(localStorage.getItem('FIRST_LOAD'))) {
      let localStrg = JSON.parse(localStorage.getItem('FIRST_LOAD'))
      setUserData(localStrg)
      setGender('All Gender')
      setInputData('')
      setHasMore(true)
    }
  }

  // Filter Gender on dropdown list
  const filterByGender = (gen) => {
    if (JSON.parse(localStorage.getItem('FIRST_LOAD'))) {
      let localStrg = JSON.parse(localStorage.getItem('FIRST_LOAD'))
      let filteredArr = localStrg.filter(val => {
        return val.gender === gen.toLowerCase()
      })
      if (filteredArr.length < 20) {
        setHasMore(false)
      }
      setUserData(filteredArr)
      setGender(gen)
    }
  }

  // Input data on search box
  const getInput = (e) => {
    setInputData(e.target.value)

    if (e.target.value === '') {
      setUserSearchData([
        {
          username: "",
          name: "",
          email: "",
          gender: "",
          registered_date: "" 
        }
      ])
    } else {
      let filteredArr = userData.filter(val => {
        return val.name.toLocaleLowerCase().includes(e.target.value.toLowerCase())
      })
      if (filteredArr.length < 20) {
        setHasMore(false)
      }
      setUserSearchData(filteredArr)
    }
  }

  // Search Data from List
  const searchData = () => {
    if (inputData.length === 0) {
      setUserSearchData([
        {
          username: "",
          name: "",
          email: "",
          gender: "",
          registered_date: "" 
        }
      ])
    } else {
      let filteredArr = userData.filter(val => {
        return val.name.toLocaleLowerCase().includes(inputData.toLowerCase())
      })
      if (filteredArr.length < 20) {
        setHasMore(false)
      }
      setUserSearchData(filteredArr)
    }
  }

  // Mapped List of Random Data
  const listData = () => {
    if (inputData.length > 0) {
      return userSearchData.map((val, ind) => {
        return (
          <tr key={ind}>
            <td>{val.username}</td>
            <td>{val.name}</td>
            <td>{val.email}</td>
            <td>{val.gender}</td>
            <td>{moment(val.registered_date).format("DD-MM-YYYY H:mm")}</td>
          </tr>
        )
      })
    } else {
      return userData.map((val, ind) => {
        return (
          <tr key={ind}>
            <td>{val.username}</td>
            <td>{val.name}</td>
            <td>{val.email}</td>
            <td>{val.gender}</td>
            <td>{moment(val.registered_date).format("DD-MM-YYYY H:mm")}</td>
          </tr>
        )
      })
    }
  }

  return (
    <div>
      <Navbar sticky='top' bg='dark' variant='dark'>
        <Container>
          <Navbar.Brand>Ajaib Random User</Navbar.Brand>
          <Navbar.Toggle />
        </Container>
      </Navbar>

      <div style={{ display: 'flex', margin: '30px' }}>
        <Form className="d-flex">
          <FormControl
            type="search"
            placeholder="Search"
            className="me-2"
            aria-label="Search"
            onChange={(e) => getInput(e)}
            value={inputData}
          />
          <Button variant="outline-success" onClick={() => searchData()}>Search</Button>
        </Form>

        <Dropdown style={{ marginLeft: '30px' }}>
          <Dropdown.Toggle variant="secondary">
           {gender}
          </Dropdown.Toggle>

          <Dropdown.Menu variant="dark">
            <Dropdown.Item onClick={() => filterByGender('Female')}>Female</Dropdown.Item>
            <Dropdown.Item onClick={() => filterByGender('Male')}>Male</Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>

        <Button variant='secondary' style={{marginLeft: '10px'}} onClick={resetFilter}>Reset Filter</Button>
      </div>

      <div style={{margin: "30px", justifyContent: 'center', alignItems: 'center'}}>
        <InfiniteScroll
          dataLength={userData.length}
          next={fetchRandomDataByPagination}
          hasMore={hasMore}
          loader={<TailSpin
            type="TailSpin"
            color="blue"
            height={100}
            width={100}
          />}
        >
        <Table striped bordered>
          <thead>
            <tr>
              <th>Username</th>
              <th>Name <FontAwesomeIcon icon={faSort} onClick={() => sortTable('name')}/></th>
              <th>Email <FontAwesomeIcon icon={faSort} onClick={() => sortTable('email')}/></th>
              <th>Gender <FontAwesomeIcon icon={faSort} onClick={() => sortTable('gender')}/></th>
              <th>Registered Date <FontAwesomeIcon icon={faSort} onClick={() => sortTable('date')}/></th>
            </tr>
          </thead>
          <tbody>
            {listData()}
          </tbody>
        </Table>
        </InfiniteScroll>
      </div>

    </div>
  );
}

export default App;
