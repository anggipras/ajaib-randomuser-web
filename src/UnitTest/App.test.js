import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import { Button, FormControl } from "react-bootstrap"
import App from "../App"

describe('App tests', function() {
    it("render search box", () => {
        const wrapper = shallow(<App />)
        expect(wrapper.find('FormControl').debug()).toEqual(
            `<FormControl type="search" placeholder="Search" className="me-2" aria-label="Search" onChange={[Function: onChange]} value="" />`
        )
    })

    it("mocking render search box", () => {
        const getInput = { value: 'Anggi'};
        const mockEvent = { target: { name: 'keyword', value: 'Anggi' } }
        const formControl = shallow((<FormControl
            type="search"
            placeholder="Search"
            className="me-2"
            aria-label="Search"
            onChange={(e) => {
                getInput[e.target.name] = e.target.value
            }}
          />));
        formControl.find("input").at(0).simulate("change", mockEvent)
        expect(getInput.value).toEqual('Anggi')
    })

    it("render reset button", () => {
        const wrapper = shallow(<App />)
        const component = `<Button variant=\"secondary\" style={{...}} onClick={[Function: resetFilter]} active={false} disabled={false}>Reset Filter</Button>`
        expect(wrapper.contains(component)).toEqual(false) //the result is false since there are 2 Button (Search and Reset Filter)
    })

    it("mocking render reset button", () => {
        const mockCallBack = jest.fn();
        const button = shallow((<Button variant='secondary' style={{marginLeft: '10px'}} onClick={mockCallBack}>Reset Filter</Button>));
        button.find('button').simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    })

    it("render correctly to matchsnapshotw", () => {
        const tree = shallow(<App />)
        expect(toJson(tree)).toMatchSnapshot()
    })
})
