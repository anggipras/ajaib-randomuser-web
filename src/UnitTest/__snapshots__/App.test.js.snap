// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests render correctly to matchsnapshot 1`] = `
<div>
  <Navbar
    bg="dark"
    collapseOnSelect={false}
    expand={true}
    sticky="top"
    variant="dark"
  >
    <Container
      fluid={false}
    >
      <NavbarBrand>
        Ajaib Random User
      </NavbarBrand>
      <NavbarToggle
        label="Toggle navigation"
      />
    </Container>
  </Navbar>
  <div
    style={
      Object {
        "display": "flex",
        "margin": "30px",
      }
    }
  >
    <Form
      className="d-flex"
    >
      <FormControl
        aria-label="Search"
        className="me-2"
        onChange={[Function]}
        placeholder="Search"
        type="search"
        value=""
      />
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="outline-success"
      >
        Search
      </Button>
    </Form>
    <Dropdown
      align="start"
      autoClose={true}
      navbar={false}
      style={
        Object {
          "marginLeft": "30px",
        }
      }
    >
      <DropdownToggle
        variant="secondary"
      >
        All Gender
      </DropdownToggle>
      <DropdownMenu
        flip={true}
        variant="dark"
      >
        <DropdownItem
          onClick={[Function]}
        >
          Female
        </DropdownItem>
        <DropdownItem
          onClick={[Function]}
        >
          Male
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "10px",
        }
      }
      variant="secondary"
    >
      Reset Filter
    </Button>
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "margin": "30px",
      }
    }
  >
    <InfiniteScroll
      dataLength={1}
      hasMore={false}
      loader={
        <TailSpin
          color="blue"
          height={100}
          type="TailSpin"
          width={100}
        />
      }
      next={[Function]}
    >
      <ForwardRef
        bordered={true}
        striped={true}
      >
        <thead>
          <tr>
            <th>
              Username
            </th>
            <th>
              Name 
              <FontAwesomeIcon
                beat={false}
                beatFade={false}
                border={false}
                bounce={false}
                className=""
                fade={false}
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [
                        "unsorted",
                      ],
                      "f0dc",
                      "M27.66 224h264.7c24.6 0 36.89-29.78 19.54-47.12l-132.3-136.8c-5.406-5.406-12.47-8.107-19.53-8.107c-7.055 0-14.09 2.701-19.45 8.107L8.119 176.9C-9.229 194.2 3.055 224 27.66 224zM292.3 288H27.66c-24.6 0-36.89 29.77-19.54 47.12l132.5 136.8C145.9 477.3 152.1 480 160 480c7.053 0 14.12-2.703 19.53-8.109l132.3-136.8C329.2 317.8 316.9 288 292.3 288z",
                    ],
                    "iconName": "sort",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                maskId={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                shake={false}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                titleId={null}
                transform={null}
              />
            </th>
            <th>
              Email 
              <FontAwesomeIcon
                beat={false}
                beatFade={false}
                border={false}
                bounce={false}
                className=""
                fade={false}
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [
                        "unsorted",
                      ],
                      "f0dc",
                      "M27.66 224h264.7c24.6 0 36.89-29.78 19.54-47.12l-132.3-136.8c-5.406-5.406-12.47-8.107-19.53-8.107c-7.055 0-14.09 2.701-19.45 8.107L8.119 176.9C-9.229 194.2 3.055 224 27.66 224zM292.3 288H27.66c-24.6 0-36.89 29.77-19.54 47.12l132.5 136.8C145.9 477.3 152.1 480 160 480c7.053 0 14.12-2.703 19.53-8.109l132.3-136.8C329.2 317.8 316.9 288 292.3 288z",
                    ],
                    "iconName": "sort",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                maskId={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                shake={false}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                titleId={null}
                transform={null}
              />
            </th>
            <th>
              Gender 
              <FontAwesomeIcon
                beat={false}
                beatFade={false}
                border={false}
                bounce={false}
                className=""
                fade={false}
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [
                        "unsorted",
                      ],
                      "f0dc",
                      "M27.66 224h264.7c24.6 0 36.89-29.78 19.54-47.12l-132.3-136.8c-5.406-5.406-12.47-8.107-19.53-8.107c-7.055 0-14.09 2.701-19.45 8.107L8.119 176.9C-9.229 194.2 3.055 224 27.66 224zM292.3 288H27.66c-24.6 0-36.89 29.77-19.54 47.12l132.5 136.8C145.9 477.3 152.1 480 160 480c7.053 0 14.12-2.703 19.53-8.109l132.3-136.8C329.2 317.8 316.9 288 292.3 288z",
                    ],
                    "iconName": "sort",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                maskId={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                shake={false}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                titleId={null}
                transform={null}
              />
            </th>
            <th>
              Registered Date 
              <FontAwesomeIcon
                beat={false}
                beatFade={false}
                border={false}
                bounce={false}
                className=""
                fade={false}
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      320,
                      512,
                      Array [
                        "unsorted",
                      ],
                      "f0dc",
                      "M27.66 224h264.7c24.6 0 36.89-29.78 19.54-47.12l-132.3-136.8c-5.406-5.406-12.47-8.107-19.53-8.107c-7.055 0-14.09 2.701-19.45 8.107L8.119 176.9C-9.229 194.2 3.055 224 27.66 224zM292.3 288H27.66c-24.6 0-36.89 29.77-19.54 47.12l132.5 136.8C145.9 477.3 152.1 480 160 480c7.053 0 14.12-2.703 19.53-8.109l132.3-136.8C329.2 317.8 316.9 288 292.3 288z",
                    ],
                    "iconName": "sort",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                maskId={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                shake={false}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                titleId={null}
                transform={null}
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="0"
          >
            <td />
            <td />
            <td />
            <td />
            <td>
              Invalid date
            </td>
          </tr>
        </tbody>
      </ForwardRef>
    </InfiniteScroll>
  </div>
</div>
`;
